File: Lab 2 - UI Non-Native/lib/model/Transaction.dart


class Transaction {

  static List<Transaction> init() {
    return [
      Transaction(1, "Transaction 1", "Income", 100.0, "12.10.2023", "Details 1"),
      Transaction(2, "Transaction 2", "Expense", 200.0, "13.10.2023", "Details 2"),
      Transaction(3, "Transaction 3", "Income", 300.0, "14.10.2023", "Details 3"),
      Transaction(4, "Transaction 4", "Expense", 400.0, "15.10.2023", "Details 4"),
      Transaction(5, "Transaction 5", "Income", 500.0, "16.10.2023", "Details 5"),

    ];
  }

  final int id;
  String name;
  String type;
  double amount;
  String date;
  String? details;

  Transaction(this.id, this.name, this.type, this.amount, this.date, this.details);
  Transaction.fromTransaction(this.id, this.name, this.type, this.amount, this.date, this.details);

  Transaction copyWith({required String name, required String type, required double amount, required String date, String? details}) {
    return Transaction.fromTransaction(this.id, name, type, amount, date, details);
  }

  @override
  String toString() {
    return "Transaction(id: $id, name: $name, type: $type, amount: $amount, date: $date, details: $details)";
  }
}

File: Lab 4 - Server Integration/SpendWiseApp/SpendWiseApp/Model/Transaction.swift
//
//  Transaction.swift
//  SpendWiseApp
//
//  Created by George Zirbo on 15.01.2024.
//

import Foundation
import SwiftData

@Model
final class Transaction: Codable, CustomStringConvertible {
    @Attribute(.unique) var id: Int
    var name: String
    var type: String
    var amount: Double
    var date: String
    var details: String

    init(id: Int, name: String, type: String, amount: Double, date: String, details: String) {
        self.id = id
        self.name = name
        self.type = type
        self.amount = amount
        self.date = date
        self.details = details
    }
    
    var description: String {
        return "Transaction(id: \(id), name: \(name), type: \(type), amount: \(amount), date: \(date), details: \(details))"
    }
    
    enum CodingKeys: CodingKey {
        case id
        case name
        case type
        case amount
        case date
        case details
    }
    
    required init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = try container.decode(Int.self, forKey: .id)
        name = try container.decode(String.self, forKey: .name)
        type = try container.decode(String.self, forKey: .type)
        amount = try container.decode(Double.self, forKey: .amount)
        date = try container.decode(String.self, forKey: .date)
        details = try container.decode(String.self, forKey: .details)
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(id, forKey: .id)
        try container.encode(name, forKey: .name)
        try container.encode(type, forKey: .type)
        try container.encode(amount, forKey: .amount)
        try container.encode(date, forKey: .date)
        try container.encode(details, forKey: .details)
    }
}
